### PR TITLE
fix(forge): do not search for optimal gas usage of `run()`

### LINKED ISSUE #3921: FFI command called multiple times
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (6ad60c8 2022-12-20T00:03:55.959118247Z)

### What command(s) is the bug in?

forge script script/poc.s.sol --ffi

### Operating System

Linux

### Describe the bug

Hi there, for weird reasons, ffi commands run from solidity scripts are executed multiple times.

Here is a PoC example:
```solidity
/* poc.s.sol */

// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.17;

import "forge-std/Script.sol";

contract RunFfi is Script {
    function run() external {
        string[] memory inputs = new string[](3);
        inputs[0] = "python3";
        inputs[1] = "script/_python/script.py";
        bytes memory res = vm.ffi(inputs);
    }
}
```

```python
# script.py

import sys
from eth_abi import encode

with open("script/_python/test.txt", "r+") as f:
    n = int(f.read())
    print(f"Call {n}", file=sys.stderr)
    f.truncate(0)
    f.seek(0)
    f.write(str(n+1))

enc = encode(["uint256"], [1])
print("0x" + enc.hex())
```

Running `forge script script/poc.s.sol --ffi ` outputs the following:
```
developer@645d22a985c4:/com.docker.devenvironments.code$ forge script script/poc.s.sol --ffi
[⠢] Compiling...
No files changed, compilation skipped
2022-12-20T17:37:22.118460Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="Call 0\n"
2022-12-20T17:37:22.229839Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="Call 1\n"
2022-12-20T17:37:22.340024Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="Call 2\n"
2022-12-20T17:37:22.452011Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="Call 3\n"
2022-12-20T17:37:22.565023Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="Call 4\n"
2022-12-20T17:37:22.673120Z ERROR foundry_evm::executor::inspector::cheatcodes::ext: stderr err="Call 5\n"
Script ran successfully.
Gas used: 29208
```

This shows that the ffi command is called 6 times which is a bit problematic when the command is expensive to run.

### PR DESCRIPTION
**Pull request by joshieDo**
closes #3921 

To search for optimal gas usage, we do need to re-execute certain calls. But as far as I can see, these should only be the ones which we intend to broadcast. No need the re-execute the `Script::run()`.

### CODE REVIEW

File: cli/src/cmd/forge/script/runner.rs [hunk 1]
```diff
@@ -278,14 +278,77 @@ impl ScriptRunner {
...
+        if commit {
+            gas_used = self.search_optimal_gas_usage(&res, from, to, &calldata, value)?;
+
+            // if we changed `fs_commit` during gas limit search, re-execute the call with original
+            // value
+            if fs_commit_changed {
+                if let Some(ref mut cheatcodes) = self.executor.inspector_config_mut().cheatcodes {
+                    cheatcodes.fs_commit = !cheatcodes.fs_commit;
+                }
+            }
```
Comment (joshieDo) [inline cli/src/cmd/forge/script/runner.rs:296]: I'm unsure of this... @mattsse

Comment (mattsse) [reply cli/src/cmd/forge/script/runner.rs:296]: this sounds like a very unlikely edge case? but hard to say without an example.

**Review by mattsse — APPROVED** [review]
makes sense, lgtm

**Comment by mattsse** [comment]
since the test changed gas calc, some files need to be updated as well, see test logs.

**Pull request merged**